### PR TITLE
Add an `is_effect` flag to `GenericArgKind`

### PR DESCRIPTION
**Pull request by fee1-dead**
Which should help us prevent printing `<false>` (effects desugaring) in diagnostics.

r? @compiler-errors 
cc @oli-obk

### CODE REVIEW

**Review by compiler-errors — CHANGES_REQUESTED** [review]
I think this approach is pretty invasive, and would rather we don't change every usage site of `GenericArg::Const` just to handle a printing bug.

I would much rather see the call to `path_generic_args` take an additional `&[ty::GenericParamDef]` so it can detect host-ness via that param instead. I think this is possible -- if you don't want to attempt this, please let me know, since I may be able to try it myself.

**Comment by fee1-dead** [comment]
> I would much rather see the call to `path_generic_args` take an additional `&[ty::GenericParamDef]` so it can detect host-ness via that param instead. I think this is possible -- if you don't want to attempt this, please let me know, since I may be able to try it myself.

Sounds like a good approach. I probably won't have more time to work on this this weekend, so feel free to pick it up if you want.

**Comment by oli-obk** [comment]
Could we add a lang item `struct Effect(bool)` that we'd use for the type of all effect consts and detect it that way without needing to change a lot?

**Comment by fee1-dead** [comment]
> Could we add a lang item `struct Effect(bool)` that we'd use for the type of all effect consts and detect it that way without needing to change a lot?

hmmm, so we can just see if the defid of the type of the const matches? sounds like a good idea too.

File: src/librustdoc/clean/utils.rs [hunk 1]
```diff
@@ -123,12 +123,16 @@ pub(crate) fn ty_args_to_args<'tcx>(
                 }),
             )))
         }
-        GenericArgKind::Const(ct) => {
+        GenericArgKind::Const(ct, is_effect) => {
             if let ty::GenericParamDefKind::Const { is_host_effect: true, .. } = params[index].kind
             {
                 return None;
             }
```
Comment (fmease) [inline src/librustdoc/clean/utils.rs:131]:
```suggestion
```

I'm pretty sure that this has become obsolete now

**Comment by oli-obk** [comment]
> hmmm, so we can just see if the defid of the type of the const matches? sounds like a good idea too.

kind of, yea. we already have the adtflags that we can check for `Box` and `UnsafeCell`, adding another entry there should not be an issue if we have the bits left to do it

**Pull request closed without merging**